packages/next/src/server/use-cache/use-cache-wrapper.ts

```diff
@@ -929,3 +916,51 @@ function shouldForceRevalidate(
 
   return false
 }
+
+function shouldDiscardCacheEntry(
+  entry: CacheEntry,
+  workStore: WorkStore,
+  implicitTags: ImplicitTags | undefined
+): boolean {
+  // If the cache entry contains revalidated tags that the cache handler might
+  // not know about yet, we need to discard it.
+  if (entry.tags.some((tag) => isRecentlyRevalidatedTag(tag, workStore))) {
+    return true
+  }
+
+  if (implicitTags) {
+    // If the cache entry was created before any of the implicit tags were
+    // revalidated last, we also need to discard it.
+    if (entry.timestamp <= implicitTags.expiration) {
+      return true
+    }
+
+    // Finally, if any of the implicit tags have been revalidated recently, we
+    // also need to discard the cache entry.
+    if (
+      implicitTags.tags.some((tag) => isRecentlyRevalidatedTag(tag, workStore))
+    ) {
+      return true
+    }
+  }
+
+  return false
+}
+
+function isRecentlyRevalidatedTag(tag: string, workStore: WorkStore): boolean {
+  const { previouslyRevalidatedTags, pendingRevalidatedTags } = workStore
+
+  // Was the tag previously revalidated (e.g. by a redirecting server action)?
+  if (previouslyRevalidatedTags.includes(tag)) {
+    return true
+  }
```

Comment:
one thing we can do is store not just the tag name but also the time at which the render started. Then instead of just blanket discarding every entry if it is tagged we make sure the timestamp is after the render start. Then when you layer on supporting in-action revalidation we would store the time of the revalidation event so we can generate a new cache entry once but then continue to use it throughout the rest of the render